File: src/lib/chatCollections.test.ts
/**
 * Exodus Browser — WebChat Collections (WebChat 收藏) API tests.
 */
import { describe, it, expect, beforeEach, vi } from 'vitest';
import {
  deriveCollectionContentType,
  collectionItemPreview,
  buildSaveChatItemRequest,
  chatCollectionSave,
  chatCollectionList,
  chatCollectionSearch,
  chatCollectionDelete,
  chatCollectionIsSaved,
} from './chatCollections';
import type { GroupMessage } from './groupChat';

vi.mock('@tauri-apps/api/core', () => ({
  invoke: vi.fn(),
  isTauri: () => true,
}));

const sampleMessage = (): GroupMessage => ({
  messageId: 'msg-1',
  groupId: 'room-1',
  senderId: 'user-a',
  senderName: 'Alice',
  content: 'See https://example.com',
  messageType: 'text',
  attachments: [],
  replyTo: null,
  mentions: [],
  timestamp: 1_700_000_000_000,
  isEdited: false,
  editedAt: null,
});

describe('chatCollections', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('derives link content type from URL', () => {
    expect(deriveCollectionContentType('visit https://exodus.dev', 'text', [])).toBe('link');
  });

  it('derives image content type from attachments', () => {
    expect(
      deriveCollectionContentType('', 'text', [
        {
          attachmentId: 'a1',
          fileType: 'image/png',
          blobHash: 'hash',
          fileName: 'photo.png',
          fileSize: 100,
        },
      ]),
    ).toBe('image');
  });

  it('builds preview text for saved items', () => {
    expect(
      collectionItemPreview({
        id: '1',
        user_id: 'u1',
        source_message_id: 'm1',
        conversation_id: 'r1',
        conversation_type: 'dm',
        conversation_title: 'Alice',
        sender_id: 'a',
        sender_name: 'Alice',
        content_type: 'text',
        content: 'Hello',
        message_type: 'text',
        attachments: [],
        original_timestamp: 1,
        saved_at: 2,
      }),
    ).toBe('Hello');
  });

  it('builds save request from group message', () => {
    const req = buildSaveChatItemRequest({
      userId: 'user-1',
      message: sampleMessage(),
      conversationType: 'dm',
      conversationTitle: 'Alice',
    });
    expect(req.sourceMessageId).toBe('msg-1');
    expect(req.conversationTitle).toBe('Alice');
  });

  it('saves collection item via invoke', async () => {
    const { invoke } = await import('@tauri-apps/api/core');
    const saved = {
      id: 'saved-1',
      user_id: 'user-1',
      source_message_id: 'msg-1',
      conversation_id: 'room-1',
      conversation_type: 'dm',
      conversation_title: 'Alice',
      sender_id: 'user-a',
      sender_name: 'Alice',
      content_type: 'link',
      content: 'See https://example.com',
      message_type: 'text',
      attachments: [],
      original_timestamp: 1_700_000_000_000,
      saved_at: 1_700_000_000_100,
    };
    vi.mocked(invoke).mockResolvedValue(saved);

    const result = await chatCollectionSave(
      buildSaveChatItemRequest({
        userId: 'user-1',
        message: sampleMessage(),
        conversationType: 'dm',
        conversationTitle: 'Alice',
      }),
    );

    expect(result).toEqual(saved);
    expect(invoke).toHaveBeenCalledWith('chat_collection_save', expect.any(Object));
  });

  it('lists collection items', async () => {
    const { invoke } = await import('@tauri-apps/api/core');
    vi.mocked(invoke).mockResolvedValue([]);
    await chatCollectionList('user-1');
    expect(invoke).toHaveBeenCalledWith('chat_collection_list', { userId: 'user-1' });
  });

  it('searches collection items', async () => {
    const { invoke } = await import('@tauri-apps/api/core');
    vi.mocked(invoke).mockResolvedValue([]);
    await chatCollectionSearch('user-1', 'hello');
    expect(invoke).toHaveBeenCalledWith('chat_collection_search', { userId: 'user-1', query: 'hello' });
  });

  it('deletes collection item', async () => {
    const { invoke } = await import('@tauri-apps/api/core');
    vi.mocked(invoke).mockResolvedValue(true);
    const ok = await chatCollectionDelete('saved-1', 'user-1');
    expect(ok).toBe(true);
    expect(invoke).toHaveBeenCalledWith('chat_collection_delete', { id: 'saved-1', userId: 'user-1' });
  });

  it('checks if message is saved', async () => {
    const { invoke } = await import('@tauri-apps/api/core');
    vi.mocked(invoke).mockResolvedValue(true);
    const saved = await chatCollectionIsSaved('user-1', 'msg-1');
    expect(saved).toBe(true);
    expect(invoke).toHaveBeenCalledWith('chat_collection_is_saved', {
      userId: 'user-1',
      sourceMessageId: 'msg-1',
    });
  });
});

File: src-tauri/src/chat_collections_test.rs
//! Unit tests for chat collections (WeChat-style 收藏).

#[cfg(test)]
mod tests {
    use super::super::chat_collections::*;
    use std::sync::Arc;
    use tempfile::TempDir;

    fn sample_request(user_id: &str, message_id: &str) -> SaveChatItemRequest {
        SaveChatItemRequest {
            user_id: user_id.to_string(),
            source_message_id: message_id.to_string(),
            conversation_id: "room-1".to_string(),
            conversation_type: "dm".to_string(),
            conversation_title: "Alice".to_string(),
            sender_id: "sender-1".to_string(),
            sender_name: "Alice".to_string(),
            content: "Hello https://example.com".to_string(),
            message_type: "text".to_string(),
            attachments: vec![],
            original_timestamp: 1_700_000_000_000,
        }
    }

    #[test]
    fn test_save_list_and_search() {
        let temp = TempDir::new().expect("temp dir");
        let path = temp.path().join("collections.json");
        let manager = ChatCollectionManager::new(path);

        let saved = manager
            .save(sample_request("user-1", "msg-1"))
            .expect("save");
        assert_eq!(saved.content_type, "link");

        let list = manager.list("user-1").expect("list");
        assert_eq!(list.len(), 1);
        assert_eq!(list[0].source_message_id, "msg-1");

        let hits = manager.search("user-1", "example").expect("search");
        assert_eq!(hits.len(), 1);

        let dup = manager.save(sample_request("user-1", "msg-1"));
        assert!(dup.is_err());
    }

    #[test]
    fn test_delete_and_is_saved() {
        let temp = TempDir::new().expect("temp dir");
        let path = temp.path().join("collections.json");
        let manager = ChatCollectionManager::new(path);

        let saved = manager
            .save(sample_request("user-1", "msg-2"))
            .expect("save");
        assert!(manager.is_saved("user-1", "msg-2").expect("is_saved"));

        let deleted = manager.delete(&saved.id, "user-1").expect("delete");
        assert!(deleted);
        assert!(!manager.is_saved("user-1", "msg-2").expect("is_saved"));
    }

    #[test]
    fn test_persistence_reload() {
        let temp = TempDir::new().expect("temp dir");
        let path = temp.path().join("collections.json");

        {
            let manager = ChatCollectionManager::new(path.clone());
            manager
                .save(sample_request("user-1", "msg-3"))
                .expect("save");
        }

        let reloaded = ChatCollectionManager::new(path);
        let list = reloaded.list("user-1").expect("list");
        assert_eq!(list.len(), 1);
        assert_eq!(list[0].source_message_id, "msg-3");
    }

    #[test]
    fn test_image_content_type() {
        let temp = TempDir::new().expect("temp dir");
        let manager = ChatCollectionManager::new(temp.path().join("collections.json"));
        let mut req = sample_request("user-1", "img-1");
        req.message_type = "image".to_string();
        req.content = String::new();
        req.attachments.push(MessageAttachmentSnapshot {
            attachment_id: "a1".to_string(),
            file_type: "image/png".to_string(),
            blob_hash: "hash".to_string(),
            file_name: "photo.png".to_string(),
            file_size: 1024,
            thumbnail_hash: None,
        });
        let saved = manager.save(req).expect("save");
        assert_eq!(saved.content_type, "image");
    }
}
